opensquirrel/utils/matrix_expander.py

```diff
@@ -0,0 +1,185 @@
+
+
+def extract_bits(x: int, bit_indices: [int]) -> int:
+    """
+    Extract the bits at given indices, placing the bits in order from least to most significant.
+    Equivalent to pext instruction.
+
+    Args:
+        x: a bitstring
+        bit_indices: the indices to extract
+
+    Returns:
+        the extracted bits of x in order
+
+    Examples:
+        >>> extract_bits(1, [0])   # 0b01
+        1
+        >>> extract_bits(1111, [2])   # 0b01
+        1
+        >>> extract_bits(1111, [5])   # 0b0
+        0
+        >>> extract_bits(1111, [2, 5])   # 0b01
+        1
+        >>> extract_bits(101, [1, 0])   # 0b10
+        2
+        >>> extract_bits(101, [0, 1])   # 0b01
+        1
+    """
+    result = 0
+    for i in range(len(bit_indices)):
+        operand = bit_indices[i]
+        result |= ((x & (1 << operand)) >> operand) << i
+
+    return result
+
+
+def deposit_bits(x: int, bit_indices: [int]) -> int:
+    """
+    Deposes the bits from x at given indices in the result.
+    Equivalent to pdep instruction.
+
+    Args:
+        x: a bitstring
+        bit_indices: the indices where to deposit the bits of x
+
+    Returns:
+        a bitstring whose bit values are taken from the bits of x
+
+    Examples:
+        >>> deposit_bits(0b0, [5])   # 0b000000
+        0
+        >>> deposit_bits(0b1, [5])   # 0b100000
+        32
+        >>> deposit_bits(0b000, [1, 2, 3])
+        0
+        >>> deposit_bits(0b001, [1, 2, 3])    # 0b0010
+        2
+        >>> deposit_bits(0b011, [1, 2, 3])    # 0b0110
+        6
+        >>> deposit_bits(0b0101, [1, 2, 3])   # 0b1010
+        10
+    """
+    result = 0
+    for i in range(len(bit_indices)):
+        result |= ((x & (1 << i)) >> i) << bit_indices[i]
+
+    return result
+
+
+def clear_bits(x: int, bit_indices: [int]) -> int:
+    """
+    Clears given bits of input.
+
+    Args:
+        x: a bitstring
+        bit_indices: bit indices to clear
+
+    Returns:
+        x with given bits reset to 0
+
+    Examples:
+        >>> clear_bits(0b1111, [1, 3])   # 0b0101
+        5
+    """
+    result = x
+    for index in bit_indices:
+        result &= ~(1 << index)
+
+    return result
+
+
+def get_expanded_matrix(semantic: Semantic, qubit_operands: [int], total_qubits: int) -> np.ndarray:
+    """
+    Get the unitary matrix corresponding to the gate applied to those qubit operands.
+    This can be used for
+    - testing,
+    - permuting the operands of a multi-qubit gates,
+    - simulating a circuit (simulation in this way is inefficient for large numbers of qubits),
+    - ...
+
+    Args:
+        semantic: the semantic of the gate
+        qubit_operands: the qubit indices on which the gate operates
+        total_qubits: the total number of qubits
+
+    Example:
+        >>> X = SingleQubitAxisAngleSemantic((1, 0, 0), math.pi, math.pi / 2)
+        >>> get_expanded_matrix(X, [1], 2).astype(int)           # X q[1]
```

Comment:
Why do you need to write `.astype(int)` here? Could this be implemented into `get_expanded_matrix` by default?

Reply:
it's ugly because it writes for each entry a real and imaginary part with 3 decimals

Reply:
I see, I general that would be the desired outcome, but not for the example here.

Reply:
indeed